rclcpp/test/rclcpp/test_shared_memory.cpp

```diff
@@ -0,0 +1,376 @@
+  EXPECT_TRUE(pub_sub_node->memory_is_reused());
+  EXPECT_EQ(pub_sub_node->get_msg_value(9), user_depth + 1);
+}
+
+TEST_F(TestSharedMemory, pub_is_last_loan_owner)
+{
+  uint16_t pub_depth = 10;
+  auto pub_sub_node = std::make_shared<PubSubNode>("pub_sub_node", pub_depth, 1, 1);
+  auto idle_node = std::make_shared<PubSubNode>("idle_node", 1, 1 ,1);
+  executor->add_node(pub_sub_node);
+
+  // Publish 11 messages. Publisher will delete oldest message (thus returning
+  // the loan)
+  pub_sub_node->borrow_and_publish_loaned(pub_depth + 1);
+  EXPECT_EQ(pub_sub_node->get_number_of_stored_messages(), 1);
+  EXPECT_FALSE(pub_sub_node->memory_is_reused());
+  EXPECT_EQ(pub_sub_node->get_msg_value(0), pub_depth);
+
+  // Publish 1 message more, memory should be reused since loan was returned
+  pub_sub_node->borrow_and_publish_loaned(1);
+  EXPECT_TRUE(pub_sub_node->memory_is_reused());
+  EXPECT_EQ(pub_sub_node->get_msg_value(0), pub_depth + 1);
+}
+
+TEST_F(TestSharedMemory, sub_is_last_loan_owner)
+{
+  uint16_t sub_depth = 10;
+  auto pub_sub_node = std::make_shared<PubSubNode>("pub_sub_node", 1, 1, 1);
+  auto idle_node = std::make_shared<PubSubNode>("idle_node", 1, sub_depth, 1);
+  executor->add_node(pub_sub_node);
+
+  // Publish 11 messages. Idle subscription will delete oldest message (thus returning
+  // the loan)
+  pub_sub_node->borrow_and_publish_loaned(sub_depth + 1);
+  EXPECT_EQ(pub_sub_node->get_number_of_stored_messages(), 1);
+  EXPECT_FALSE(pub_sub_node->memory_is_reused());
+  EXPECT_EQ(pub_sub_node->get_msg_value(0), sub_depth);
+
+  // Publish 1 message more, memory should be reused since loan was returned
+  pub_sub_node->borrow_and_publish_loaned(1);
+  EXPECT_TRUE(pub_sub_node->memory_is_reused());
+  EXPECT_EQ(pub_sub_node->get_msg_value(0), sub_depth + 1);
+}
+
+TEST_F(TestSharedMemory, memory_is_shared)
+{
+  // Publish 10 messages then overriden by the user.
+  // Idle subscription is then executed: it should receive the modified messages.
+  auto pub_sub_node = std::make_shared<PubSubNode>("pub_sub_node", 15, 1, 15);
+  auto late_joiner_node = std::make_shared<PubSubNode>("late_joiner_node", 1, 10 , 15);
+  executor->add_node(pub_sub_node);
+
+  pub_sub_node->borrow_and_publish_loaned(10);
+  pub_sub_node->user_override_messages(42);
+  EXPECT_EQ(pub_sub_node->get_msg_value(9), 42);
+
+  // Execute subscription which had the messages stored in its queue
+  executor->add_node(late_joiner_node);
+  std::this_thread::sleep_for(1s);
+
+  EXPECT_EQ(late_joiner_node->get_number_of_stored_messages(), 10);
+  // Verify it has the modified version
+  EXPECT_EQ(late_joiner_node->get_msg_value(9), 42);
+
+  auto pub_sub_loaned = pub_sub_node->get_user_messages();
+  auto sub_loaned = late_joiner_node->get_user_messages();
+  EXPECT_TRUE(loaned_addresses_match(pub_sub_loaned, sub_loaned));
+}
+
+// FastDDS: Show that the publisher gets loans from memory still in use:
```

Comment:
It looks like there are some tests that are specific to some RMW implementation.
I'm also not sure if they are all valid unit-tests or rather "examples" showing the limitations of some RMWs.

In any case, we should run these tests only for the RMWs they matter, for example with
```
  if (std::string(rmw_get_implementation_identifier()).find("rmw_connextdds") == 0)
  {
    GTEST_SKIP();
  }
```